src/flatpak-spawn.c

```diff
@@ -496,8 +498,133 @@ check_portal_supports (const char *option, guint32 supports_needed)
     }
 }
 
+/*
+ * @str: A path
+ * @prefix: A possible prefix
+ *
+ * The same as flatpak_has_path_prefix(), but instead of a boolean,
+ * return the part of @str after @prefix (non-%NULL but possibly empty)
+ * if @str has prefix @prefix, or %NULL if it does not.
+ *
+ * Returns: (nullable) (transfer none): the part of @str after @prefix,
+ *  or %NULL if @str is not below @prefix
+ */
+static const char *
+get_path_after (const char *str,
+                const char *prefix)
+{
+  while (TRUE)
+    {
+      /* Skip consecutive slashes to reach next path
+         element */
+      while (*str == '/')
+        str++;
+      while (*prefix == '/')
+        prefix++;
+
+      /* No more prefix path elements? Done! */
+      if (*prefix == 0)
+        return str;
+
+      /* Compare path element */
+      while (*prefix != 0 && *prefix != '/')
+        {
+          if (*str != *prefix)
+            return NULL;
+          str++;
+          prefix++;
+        }
+
+      /* Matched prefix path element,
+         must be entire str path element */
+      if (*str != '/' && *str != 0)
+        return NULL;
+    }
+}
+
+static gint32
+path_to_handle (GUnixFDList *fd_list,
+                const char *path,
+                const char *home_realpath,
+                const char *flatpak_id,
+                GError **error)
+{
+  int path_fd = open (path, O_PATH|O_CLOEXEC|O_NOFOLLOW|O_RDONLY);
+  int saved_errno;
+  gint32 handle;
+
+  if (path_fd < 0)
+    {
+      saved_errno = errno;
+      g_set_error (error, G_IO_ERROR, g_io_error_from_errno (saved_errno),
+                   "Failed to open %s to expose in sandbox: %s",
+                   path, g_strerror (saved_errno));
+      return -1;
+    }
+
+  if (home_realpath != NULL && flatpak_id != NULL)
+    {
+      g_autofree char *real = NULL;
+      const char *after = NULL;
+
+      real = realpath (path, NULL);
+
+      if (real != NULL)
+        after = get_path_after (real, home_realpath);
+
+      if (after != NULL)
+        {
+          g_autofree char *var_path = NULL;
+          int var_fd = -1;
+          struct stat path_buf;
+          struct stat var_buf;
+
+          /* @after is possibly "", but that's OK: if @path is exactly $HOME,
+           * we want to check whether it's the same file as
+           * ~/.var/app/$FLATPAK_ID, with no suffix */
```

Comment:
Pet peeve: I have comments that don't line up the closing */ with the rest of the *s